llvm/lib/Target/AMDGPU/AMDGPUSplitKernelArguments.cpp

```diff
@@ -0,0 +1,373 @@
+//===--- AMDGPUSplitKernelArguments.cpp - Split kernel arguments ----------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// \file This pass flats struct-type kernel arguments. It eliminates unused
+// fields and only keeps used fields. The objective is to facilitate preloading
+// of kernel arguments by later passes.
```

Comment:
Is it really worth having an ABI breaking pass for this? Why can't the use just reorder the arguments? 

This needs substantial unit test coverage, preferably in the llvm-test-suite 

I also think this is leaning too heavily on the runtime dispatch inherited from OpenCL, really we ought to be able to emit directly callable signatures without going through the runtime metadata to set up the dispatch arguments

Reply:
add a test to llvm-test-suite: https://github.com/llvm/llvm-test-suite/pull/230